src/Microsoft.OData.Core/UriParser/TypePromotionUtils.cs

```diff
@@ -495,6 +495,12 @@ internal static bool CanConvertTo(SingleValueNode sourceNodeOrNull, IEdmTypeRefe
 
             if (targetReference.IsODataComplexTypeKind() || targetReference.IsODataEntityTypeKind())
             {
+                // https://github.com/OData/odata.net/issues/1395
+                // Both targetReference and sourceReference are cast to IEdmStructuredType,
+                // therefore we must check both types
+                if (!sourceReference.IsODataComplexTypeKind() && !sourceReference.IsODataEntityTypeKind())
+                    return false;

```

Comment:
We also need to support conversions to/from untyped.  Prior to change, if targetReference is complex or entity and sourceReference.IsUntyped() is true, then CanConvertTo would return true, but with your change would return false.  Simplest is probably to add a check before 495 that returns true if targetReference.IsUntype() || sourceReference.IsUntyped().  Please also add corresponding tests for untyped.  Thanks!

Reply:
Thank you for the comment. I added the IsUntyped() check and the unit tests.

Reply:
:thumbs up: 

---
In reply to: [320697077](https://github.com/OData/odata.net/pull/1489#discussion_r320697077) [](ancestors = 320697077)